File: 917. Reverse Only Letters/_917.py
class Solution:
    def reverseOnlyLetters(self, S: str) -> str:
        res = ['a']*len(S)
        pre = 0
        lst = len(S)-1
        while pre<=lst:
            flag = 0
            if not S[pre].isalpha():
                res[pre] = S[pre]
                pre+=1
                flag = 1
            if not S[lst].isalpha():
                res[lst] = S[lst]
                lst -= 1
                flag = 1
            
            if flag == 1:
                continue

            res[pre], res[lst] = S[lst], S[pre]
            pre += 1
            lst -= 1

        return ''.join(res)
    
    def reverseOnlyLetters1(self, S):
        S, i, j = list(S), 0, len(S) - 1
        while i < j:
            if not S[i].isalpha():
                i += 1
            elif not S[j].isalpha():
                j -= 1
            else:
                S[i], S[j] = S[j], S[i]
                i, j = i + 1, j - 1
        return "".join(S)

    def reverseOnlyLetters2(self, S):
        r = [s for s in S if s.isalpha()]
        return "".join(S[i] if not S[i].isalpha() else r.pop() for i in range(len(S)))

a = Solution()
print(a.reverseOnlyLetters("a-bC-dEf-ghIj"))

File: 917. Reverse Only Letters/917.go
package leetcode

import (
	"unicode"
	"strings"
)

func reverseOnlyLetters(S string) string {
	res := make([]string, len(S))
	pre := 0
	lst := len(S)-1

	for pre<=lst{
		flag := 0
		if !unicode.IsLetter(rune(S[pre])){
			res[pre] = string(S[pre])
			pre++
			flag = 1
		}
		if !unicode.IsLetter(rune(S[lst])){
			res[lst] = string(S[lst])
			lst--
			flag = 1
		}

		if flag == 1{
			continue
		}

		res[pre], res[lst] = string(S[lst]), string(S[pre])
		pre ++
		lst --
	}
	return strings.Join(res, "")
}

func reverseOnlyLetters1(S string) string {
    start, end := 0, len(S) - 1
    ans := []byte(S)
    for start < end {
        for !isLetter(S[start]) && start < end{
            start++
        }
        for !isLetter(S[end]) && start < end {
            end--
        }
        ans[start], ans[end] = ans[end], ans[start]
        start++
        end--
    }
    return string(ans)
}

func isLetter(c byte) bool {
    return c >= 'a' && c <= 'z' || c >= 'A' && c <= 'Z'
}